lib/spack/spack/test/stage.py

```diff
@@ -177,82 +194,32 @@ def get_stage_path(stage, stage_name):
-    """Temporarily ensure there is no accessible path for staging."""
-    current = spack.config.get('config:build_stage')
-    spack.config.set('config', {'build_stage': '/no/such/path'}, scope='user')
-    yield
-    spack.config.set('config', {'build_stage': current}, scope='user')
-
-
-@pytest.fixture
-def non_user_path_for_stage(monkeypatch):
-    """Temporarily use a Linux-standard non-user path for staging. """
-    def _can_access(path, perms):
-        return True
-
-    current = spack.config.get('config:build_stage')
-    spack.config.set('config', {'build_stage': [_non_user_root]}, scope='user')
-    monkeypatch.setattr(os, 'access', _can_access)
-    yield
-    spack.config.set('config', {'build_stage': current}, scope='user')
-
-
-@pytest.fixture
-def instance_path_for_stage():
-    """
-    Temporarily use the "traditional" spack instance stage path for staging.
-
-    Note that it can be important for other tests that the previous settings be
-    restored when the test case is over.
-    """
-    current = spack.config.get('config:build_stage')
+@pytest.mark.nomockstage
+def instance_path_for_stage(clear_stage_root):
+    """Use the spack instance stage path for the stage root."""
     base = canonicalize_path(os.path.join('$spack', 'test-stage'))
     mkdirp(base)
-    path = tempfile.mkdtemp(dir=base)
-    spack.config.set('config', {'build_stage': path}, scope='user')
-    yield
-    spack.config.set('config', {'build_stage': current}, scope='user')
-    shutil.rmtree(base)
+    test_path = tempfile.mkdtemp(dir=base)
 
+    with spack.config.override('config:build_stage',  test_path):
+        yield spack.stage.get_stage_root()
 
-@pytest.fixture
-def tmp_path_for_stage(tmpdir):
-    """
-    Use a temporary test directory for staging.
-
-    Note that it can be important for other tests that the previous settings be
-    restored when the test case is over.
-    """
-    current = spack.config.get('config:build_stage')
-    spack.config.set('config', {'build_stage': [str(tmpdir)]}, scope='user')
-    yield tmpdir
-    spack.config.set('config', {'build_stage': current}, scope='user')
+    shutil.rmtree(base)
 
 
 @pytest.fixture
-def tmp_build_stage_dir(tmpdir):
-    """Establish the temporary build_stage for the mock archive."""
-    test_stage_path = tmpdir.join('stage')
+def tmp_build_stage_dir(tmpdir, clear_stage_root):
+    """Use a temporary test directory for the stage root."""
+    test_path = str(tmpdir.join('stage'))
+    with spack.config.override('config:build_stage', test_path):
+        yield tmpdir, spack.stage.get_stage_root()
 
-    # Set test_stage_path as the default directory to use for test stages.
-    current = spack.config.get('config:build_stage')
-    spack.config.set('config',
-                     {'build_stage': [str(test_stage_path)]}, scope='user')
-
-    yield (tmpdir, test_stage_path)
-
-    spack.config.set('config', {'build_stage': current}, scope='user')
+    shutil.rmtree(test_path)
```

Comment:
Is this necessary inside the `tmpdir` fixture?

Reply:
I suppose not though I am inclined to cleanup after tests.

Reply:
I would have expected the `tmpdir` pytest fixture to automatically clean up the directory (this is distinct from the concept of a filesystem temporary directory); if you observe that usage of the `tmpdir` fixture leaves files in the filesystem temp directory then we ought to figure out how to update the `tmpdir` fixture to avoid that (i.e. wrap it with our own fixture).

Reply:
The docs state `tmpdir` "will provide a temporary directory unique to the test invocation, created in the base temporary directory."  (See http://doc.pytest.org/en/latest/tmpdir.html#the-tmpdir-fixture.)

The directory does not get removed.

Wrapping `tmpdir` sounds like something outside the scope of this PR.

Reply:
```
The base name will be pytest-NUM where NUM will be incremented with each test run. Moreover, entries older than 3 temporary directories will be removed.
```

From the pytest docs. They do garbage collection on tmpdir.

I think we're right all around. They do get cleaned up, it's probably safer if we clean up ourselves anyway, and we could write a fixture to automate that, but it's outside the scope of this PR.